### PR TITLE
Add routes command to flask cli

### PR DESCRIPTION
**Pull request by willsoto**
Running `flask routes` will list all registered routes for a given app, sorted by endpoint

Inspired by [`mix phoenix.routes`](http://www.phoenixframework.org/docs/routing), this will display a list of all registered routes in an application. This is basically #1446, but I have updated it and made it "dumber". 

NB: I know that the other PR is old and there doesn't seem to be much enthusiasm for it, but I do think this is useful to have to get a high level overview of application routes, especially if they are being generated programmatically. 

Example output:

```sh
$ flask routes
api.me                    GET              /api/users/me/
api.users                 GET              /api/users/
                          POST             /api/users/
                          PUT, DELETE, GET /api/users/<pk>
api.users_schema          GET              /api/users/schema/
auth_blueprint.authorized GET              /login/authorized
auth_blueprint.do_login   GET              /do-login
auth_blueprint.login      GET              /login
auth_blueprint.logout     GET              /logout
auth_blueprint.lr_test    GET              /lr
static                    GET              /static/<path:filename>
```

### CODE REVIEW

File: flask/cli.py [hunk 1]
```diff
@@ -466,6 +467,36 @@ def shell_command():
...
+    """Prints all the given routes for an application"""
+    from flask.globals import _app_ctx_stack
+
+    app = _app_ctx_stack.top.app
+    routes = []
+    ignored_methods = set(['HEAD', 'OPTIONS'])
+    sorted_rules = sorted(app.url_map.iter_rules(), key=lambda rule: rule.endpoint)
+
+    for rule in sorted_rules:
+        methods = ', '.join(rule.methods.difference(ignored_methods))
```
Comment (ThiefMaster) [inline flask/cli.py]: I'd use the `-` operator instead of the `difference` method

**Comment by AlJohri** [comment]
👍

**Comment by paradox41** [comment]
@ThiefMaster done. Do you know what is going on with all the failing tests?

File: tests/test_cli.py [hunk 2]
```diff
@@ -190,3 +190,41 @@ def test():
...
+    def cli(**params):
+        pass
+
+    runner = CliRunner()
+    result = runner.invoke(cli, ['routes'])
+
+    output = [line.strip() for line in result.output.split('\n')]
+
+    assert result.exit_code == 0
+    assert output[0] == 'routes_blueprint.posts_routes  PUT, DELETE, GET /posts/<post_id>'
```
Comment (ThiefMaster) [inline tests/test_cli.py]: You aren't sorting the list of methods so you can't assume any specific order on data coming from a set.

File: tests/test_cli.py [hunk 3]
```diff
@@ -190,3 +190,41 @@ def test():
...
+    assert output[1] == 'routes_blueprint.user_id_route DELETE, GET      /users/<user_id>'
```
Comment (ThiefMaster) [inline tests/test_cli.py]: same

File: flask/cli.py [hunk 4]
```diff
@@ -466,6 +467,37 @@ def shell_command():
...
+@click.command('routes', short_help='Show application routes.')
+@with_appcontext
+def routes_command():
+    """Prints all the given routes for an application"""
+    from flask.globals import _app_ctx_stack
+
+    app = _app_ctx_stack.top.app
+    routes = []
+    ignored_methods = set(['HEAD', 'OPTIONS'])
+    sorted_rules = sorted(app.url_map.iter_rules(), key=lambda rule: rule.endpoint)
```
Comment (ThiefMaster) [inline flask/cli.py]: `key=attrgetter('endpoint')`

**Comment by paradox41** [comment]
@ThiefMaster Thanks for feedback. I've made all the changes you requested.

**Comment by davidism** [comment]
Thanks for doing this. I started working on yours then decided I liked the output in #1446 better, so I reworked it with some of your ideas in #2259.

**Pull request closed without merging**